### ISSUE TITLE
[BUG] `issue-only-any-project-cards` doesn't find any linked project cards

### ISSUE THREAD

**Issue by iainlane** [Sonia-corporation/stale#498]
**Describe the bug**

I see this output when running the action (1.46.0) with `issue-only-any-project-cards` set to the name of my project:

```
The issue #25344 can be processed
[#25344] Processing the issue 25344 (​[https://github.com/XXX/YYY/issues/25344​)](https://github.com/XXX/YYY/issues/25344%E2%80%8B))...
  [#[253](https://github.com/XXX/YYY/runs/5199822071?check_suite_focus=true#step:4:253)44] Checking if this issue should be ignored...
  [#25344] Checking if this issue is locked...
  [#25344] Not locked. Continuing...
  [#25344] Checking if all the labels on this issue should be ignored...
  [#25344] The input issue-ignore-all-labels is disabled. Continuing...
  [#25344] Checking if this issue has one of the ignored labels...
  [#25344] Not containing an ignored label. Continuing...
  [#25344] Checking if all the assignees on this issue should be ignored...
  [#25344] The input issue-ignore-all-assignees is disabled. Continuing...
  [#25344] Checking if this issue has one of the ignored assignees...
  [#25344] Not containing an ignored assignee. Continuing...
  [#25344] Checking if all the project cards on this issue should be ignored...
  [#25344] The input issue-ignore-all-project-cards is disabled. Continuing...
  [#25344] Checking if this issue should be ignored based on its creation date...
  [#25344] The input issue-ignore-before-creation-date is either unset or not convertible to a valid ISO 8601 date. Continuing...
  [#25344] Checking if this issue should be included in the processing...
  [#25344] Checking if this issue should only be processed based on any of the associated project cards...
  [#25344] The input issue-only-any-project-cards is set. This feature is considered as enabled, and so, may alter the processing. Checking...
  [#25344] Not containing any project card. Skipping the processing of this issue...
  [#25344] Ignored
  Ignored issues count statistic increased by 1 (6 in total)
  [#25344] Processing stopped
```

The line:

> Not containing any project card. Skipping the processing of this issue...

is wrong: the issue has a project card: see this screenshot (sorry it is almost all blacked out!)

<img width="1279" alt="image" src="https://user-images.githubusercontent.com/321014/154063005-4b2b8bdc-20a7-4998-af75-7922b84b772f.png">

**Expected behaviour**
The issue should not have been ignored. The project card should have been found, and then the name of the associated project should have matched the one I gave in the config.

This particular issue is too new, so it still shouldn't have been commented on / closed, though.

**Additional context**

**Configuration**

```
      - name: Stale (XYXY)
        uses: Sonia-corporation/stale@1.46.0
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          issue-days-before-stale: 30
          issue-days-before-close: 7
          pull-request-days-before-stale: 30
          pull-request-days-before-close: 7
          # Issues and PRs with these labels will never be considered stale
          issue-ignore-any-labels: |
            pinned
            security
            type/security
            keepalive
          pull-request-ignore-any-labels: |
            pinned
            security
            type/security
            keepalive
          issue-stale-comment: >
            This issue has been automatically marked as stale because it has not
            had recent activity. It will be closed if no further activity
            occurs. Thank you for your contribution.
          pull-request-stale-comment: >
            This pull request has been automatically marked as stale because it
            has not had recent activity. It will be closed if no further
            activity occurs. Thank you for your contribution.
          # Only touch XXX / YYY issues/prs
          issue-only-any-project-cards: |
            XXX
            YYY
          pull-request-only-any-project-cards: |
            XXX
            YYY
          # While we're checking whether this works, run it as a dry-run
          dry-run: true
```

**Comment by iainlane** [comment]
btw, I think you'd expect this, but just to confirm: I do also have a run with `issue-ignore-all-project-cards: true`, and that fails to find the project card also-

```
  [#25344] The input issue-ignore-all-project-cards is enabled. Checking...
  [#25344] The issue has no project card. Continuing...
```

**Comment by C0ZEN** [comment]
Using the https://docs.github.com/en/graphql/overview/explorer, I can tell so far that the query seems well written.
I am able to run it on this repository and have the values as expected for the projects.

```graphql
query GetIssues($owner: String!, $repository: String!, $issuesPerPage: Int!, $labelsPerIssue: Int!, $assigneesPerIssue: Int!, $projectCardsPerIssue: Int!, $afterCursor: String) {
  repository(name: $repository, owner: $owner) {
    issues(
      orderBy: {field: UPDATED_AT, direction: DESC}
      states: OPEN
      first: $issuesPerPage
      after: $afterCursor
    ) {
      pageInfo {
        endCursor
        hasNextPage
      }
      totalCount
      nodes {
        locked
        createdAt
        number
        updatedAt
        url
        id
        labels(first: $labelsPerIssue) {
          totalCount
          nodes {
            id
            name
          }
        }
        assignees(first: $assigneesPerIssue) {
          totalCount
          nodes {
            login
          }
        }
        projectCards(first: $projectCardsPerIssue) {
          totalCount
          nodes {
            id
            project {
              name
            }
          }
        }
      }
    }
  }
}
```

```json
{
  "owner": "Sonia-corporation",
  "repository": "stale",
  "issuesPerPage": 20,
  "labelsPerIssue": 20,
  "assigneesPerIssue": 20,
  "projectCardsPerIssue": 20
}
```

**Comment by C0ZEN** [comment]
I am not sure about one thing: maybe the given GitHub token should have explicit write permission on your side?
See https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token

I was able to make this feature work on this repository on this run: https://github.com/Sonia-corporation/stale/runs/5203890740?check_suite_focus=true.
I will add more logs so that you will know what are the associated project cards for each issue in the logs, but I was not able to spot a bug so far.

**Comment by iainlane** [comment]
mmm, when I run the query myself (using my own account though), I can see the project card.

I guess it'd be good to basically dump out the raw data we get back from graphql, maybe under `ACTIONS_STEP_DEBUG`?

**Comment by C0ZEN** [comment]
No, the log is related to the search.
The code is trying to find a match between the list you have, and the list associated to the issue.
Well, at least for the `issue-ignore-all-project-cards` input.

Sadly, I won't have the time to merge a new PR today to bring more logs, but tomorrow should be fine.

**Comment by C0ZEN** [comment]
I created a new release that:
- expose the whole item (issue or pr) into a debug log; this is the raw data coming from GitHub GraphQL
- add a log that list the project names found before trying to find if there is a match with the white-listed project names

You will be able to tell me if in your case, you have the data as expected.
If this is not the case, it may simply be an issue of rights related to your token.
If this is good, and you can simply see that there is no match, it may be related to a comparison issue between the GitHub project name and the input name.
This could occur for numerous reasons, but I am thinking about different kind of spaces in particular.

**Comment by C0ZEN** [comment]
See https://github.com/Sonia-corporation/stale/releases/tag/1.47.0

**Comment by iainlane** [comment]
Awesome, thanks, that helped a lot!

> If this is not the case, it may simply be an issue of rights related to your token.

It is this indeed. As far as I can tell, the `GITHUB_TOKEN` which is auto-generated for workflow executions does not have `org:read` permissions. I switched to using a PAT and now the action says:

```
  [#25344] Found 1 project card on this issue XXX
  [#25344] Containing one of the required project card -> XXX
  [#25344] Continuing the processing for this issue...
```

which sounds right!

So all good, thanks for adding the debugging, it helped because I could see that the API was returning empty lists.

Perhaps this is something which could be documented? To use these `project` configuration entries, you probably need to use a different token for the API.

**Comment by C0ZEN** [comment]
@iainlane thanks for being the guinea pig on this one.
I will add a troubleshooting page to address this kind of particular case.
